Add basic search with Fuse.js (search engine), Mark.js (word highlighter) and Hugo (search index)

kinow opened the pull request:
Had some 40 minutes this morning while waiting to go out for a local party (st eulalia with giants and castellers! :partying_face:), so had a go at adding search to Jena's website.

Hugo has a page to [Search for your Hugo Website](https://gohugo.io/tools/search/) with links go gists, libraries, and posts. After some searching, I found this one to be up to date and very simple to follow: https://makewithhugo.com/add-search-to-a-hugo-site/ 

It uses Fuse.js, a fast and dependency-free JS library (really rare!) to perform full text search with support to fuzzy searching. The post also uses Mark.js, which I am not sure if was really necessary, as it's a library to highlight words, but I preferred to just add it for now and decide to drop it or not later.

The layout of the search results was not customized, we can change how they are displayed later if needed. I focused more on the search input field/form for now (due to short time). See screenshots below where I searched for "infrenc":

![image](https://user-images.githubusercontent.com/304786/218301142-660b6bd4-267b-4152-ba90-b0f3cd46692c.png)

![image](https://user-images.githubusercontent.com/304786/218301165-795f3847-87ef-4f87-96be-32e7de04e29a.png)

And, of course, if you search for a correct term, then it's highlighted in the results, like "query builder":

![image](https://user-images.githubusercontent.com/304786/218301220-eb814c85-8753-4c09-90b5-07fd001116c7.png)

That's it :+1: 
Cheers,

-Bruno

kinow commented:
p.s. we could also use Algolia free, which I use for some sites I maintain, but someone needs to have the account and maintain the token, and it has a high limit but that Jena could reach as we have a lot of users I think... so in the end I thought this was the simplest option to have search in our site, for now :+1: 

Ah! And here's the search on smaller viewports.

![image](https://user-images.githubusercontent.com/304786/218301276-bf0ebf03-ab38-4aff-8585-b5f1c48cf5ed.png)

![image](https://user-images.githubusercontent.com/304786/218301283-84a6bc10-3bc1-4aba-9cd8-e47adedcef00.png)

afs commented:
> so in the end I thought this was the simplest option to have search in our site, for now

+1

afs commented:
I pulled the branch and ran hugo.

When the screen size is a bit bigger than small the alignment goes wonky:

![Screenshot from 2023-02-12 11-48-03](https://user-images.githubusercontent.com/101362/218309404-9dc7f0c2-b04f-4388-bda2-4fd13e3c54c9.png)

I searched for "SHACL" (browser = Firefox) but it stays on "Loading ..."

kinow commented:
>I searched for "SHACL" (browser = Firefox) but it stays on "Loading ..."

I just looked at my local working copy, and I did a `git commit -am ...` and then left, without noticing that I forgot to add two new files :laughing: sorry.

New commit will push these missing files and this Loading... should either disappear very quickly or not even be visible (running locally Fuse.js renders so fast I cannot see the message).

kinow commented:
>When the screen size is a bit bigger than small the alignment goes wonky:

Oh, that happens if I remove the search field too :disappointed_relieved: let me take a look and see if I can find a solution for this...

![image](https://user-images.githubusercontent.com/304786/218314110-6c89a06b-3bfc-4067-a68e-4718ebe2cf29.png)

afs commented:
I get results!
A lot of results :-(
Each 3 times - hits first then what seems to be everything.

![image](https://user-images.githubusercontent.com/101362/218314552-52a35a74-fd6f-4e81-8b95-bb332e2e9482.png)

... it goes on a lot further - it's croipped by the screen grab.

kinow commented:
>I get results!

Hooray! :sun_with_face: 

>A lot of results :-(

Hoo... ray? :sun_behind_small_cloud: 

>Each 3 times - hits first then what seems to be everything.

Huh :sun_behind_rain_cloud: 

I see what you mean. I was happy to get results, and didn't try to tweak the Fuse.js nor Hugo index settings. For the duplicates I think I already found a solution (include an ID in the returned values). I'm going through the Fuse.js options now to tweak it a bit more...

kinow commented:
Still working on the top menu responsiveness when we resize the screen.

But I think the search results should be:

- unique (did a `| uniq`) in Hugo when it creates the index, so no more duplicates :crossed_fingers:, and
- removing some pages, like the Search results page :sweat_smile: (noticed that while testing it, fixed by adding a front matter option `excludeSearch`).

Also removed options that did not appear in the Fuse.js docs (probably the author of the blog was using an older version previously to writing it), and added comments to the ones we are using, as well as a link to the Fuse.js docs—in case anyone needs to tweak it.

kinow commented:
Now testing why some results like "TDB Transactions" are appearing in the results, as I cannot find SHACL in that page :thinking:

kinow commented on layouts/_default/search.html:
```diff
@@ -0,0 +1,200 @@
+        //       (default 0). Since Jena docs may have very long text that includes the query term anywhere
+        //       we disable it with ignoreLocation: true.
+        //       For more: https://fusejs.io/concepts/scoring-theory.html#scoring-theory
+        ignoreLocation: true,
```
@afs Fuse.js uses the location of the match in its algorithm, which IMO doesn't make much sense for our use case. For example, by default it excludes documents that have the search match appearing 60 after the initial 60 characters.

The setting above disables it.

kinow commented on layouts/_default/search.html:
```diff
@@ -0,0 +1,200 @@
+        // At what point does the match algorithm give up.
+        // A threshold of 0.0 requires a perfect match (of both letters and location),
+        // a threshold of 1.0 would match anything.
+        threshold: 0.2,
```
With a `threshold` of `0` I get `7` search results for SHACL. That's the same number I get when grepping for it (case insensitive),

```bash
kinow@ranma:~/Development/java/jena/jena-site/source$ grep -r -H -o -i SHACL | awk -F: '{ print $1 }' | sort -h | uniq
documentation/fuseki2/fuseki-config-endpoint.md
documentation/__index.md
documentation/javadoc.md
documentation/notes/system-initialization.md
documentation/shacl/__index.md
documentation/tools/__index.md
download/maven.md
```

But if I search for "shakl" it brings `0` results.

With `0.2`, both SHACL and SHAKL bring me 14 search results. The 7 first results have a score lower than `1` (in Fuse.js higher is worse), and the other 7 have a score of `1` (I left the score to be displayed with results to help users).

So I decided to leave it to 0.2 so users still get some result if they misspell their search query.

kinow commented:
I enabled the extended search of Fuse.js, that allows exact terms, and search for something that starts with a term, or ends with, or excludes a term... but that didn't work quite well. It was searching the document title, I think? :disappointed_relieved: 

And the syntax was different than what users of Google/DDG would expect (e.g. instead of "term" for an exact match, it was "=term"). So I didn't want our readers to get familiar/used to that syntax, in case we ever need to switch JS search libraries :+1: 

Ready for another review :crossed_fingers:

leeoniya commented:
> in case we ever need to switch JS search libraries

this could save you some time: https://github.com/leeoniya/uFuzzy#a-biased-appraisal-of-similar-work

kinow commented:
>this could save you some time: https://github.com/leeoniya/uFuzzy#a-biased-appraisal-of-similar-work

Thanks @leeoniya. Will keep that in mind if we have issues with Fuse.js and need to pick another library. For the moment Fuse.js seems to be OK :+1:

kinow commented:
(rebased)

afs commented:
I don't think "staging" works. They were pre-Hugo.

If it merged to main, then it's jena.apache.org - I see
7905e130d Add basic search with Fuse.js ... in git.

On Jenkins there is

https://ci-builds.apache.org/job/Jena/job/Jena_Site2/job/main/213/console

which has a build failure.

Error: Error building site: failed to render pages: render of "home" failed: "/home/jenkins/712657a4/workspace/Jena_Jena_Site2_main/layouts/_default/index.json:7:28": execute of template failed: template: _default/index.json:7:28: executing "_default/index.json" at <uniq>: error calling uniq: elements must be comparable

which I do not understand.
I tried again - same thing.

https://ci-builds.apache.org/job/Jena/job/Jena_Site2/job/main/

My hugo is hugo v0.111.3

hugo version is the Jenkinsfile at top of jena-site:main is 0.66.0
so my guess is hugo version.

kinow commented:
>hugo version is the Jenkinsfile at top of jena-site:main is 0.66.0
>so my guess is hugo version.

Yeah, searched for issues, but did not find anything recent (what I found was from ~2019).

My Hugo version:

```bash
hugo version
hugo v0.111.3-5d4eb5154e1fed125ca8e9b5a0315c4180dab192+extended linux/amd64 BuildDate=2023-03-12T11:40:50Z VendorInfo=snap:0.111.3
```

Should I open a ticket to Infra to ask them to update the Hugo version of the Jenkins workers?

afs commented:
Hugo isn't installed on the Jenkins nodes - it's pulled each time.

This is controlled by [jena-site/Jenkinsfile](https://github.com/apache/jena-site/blob/main/Jenkinsfile).

The version number is [Jenkinsfile#L26](https://github.com/apache/jena-site/blob/dd32b8a44ec71ba8bc94424da9e55d2b82b5f12e/Jenkinsfile#L26) and hugo is downloaded at [Jenkinsfile#L43](https://github.com/apache/jena-site/blob/dd32b8a44ec71ba8bc94424da9e55d2b82b5f12e/Jenkinsfile#L43).

Edit and commit!

kinow commented:
That fixed it, @afs. Thanks!!!
